### PR TITLE
Add metrics

### PR DESCRIPTION
**Pull request by hoolioh**
# What does this PR do?

Adds support for sending the following metrics:

- trace_api.bytes
- trace_chunks_sent
- trace_chunks_dropped

Aside from the introduction of those new metrics a small refactor has been done on send_data.rs in order to address some technical debt from previous PRs. This changes can be summarized in:

- Limit visibility of `SendData` fields.
- Add default trait implementation for `SendDataResult`.
- Add tests for both `TraceHeaderTags` and `SendData`.

# Motivation

In order to turn the sidecar on by default for PHP there is the need to support those metrics.

### CODE REVIEW

File: trace-utils/src/send_data.rs
Comment (pierotibou) [inline trace-utils/src/send_data.rs:252]: You should remove it I assume.

File: trace-utils/src/send_data.rs
Comment (pierotibou) [inline trace-utils/src/send_data.rs:286]: nitpicking - Ignore for this PR, but I'm not a big fan of inlined string. I believe that it's better to define constants somewhere. It's also easier to document what they are meant for.

**Review by pierotibou — COMMENTED** [review]
Thanks a lot for doing this. TLDR, I assume we should comment the sending of the `bytes` metrics (we can keep the rest) as it's not yet supported. And add a few missing `s` to metric names. 
The rest is mainly questions for my knowledge and some nitpicks that you should ignore :p

File: trace-utils/src/send_data.rs
Comment (hoolioh) [reply trace-utils/src/send_data.rs:23]: As Bob said that `enum` is to track the errors coming from the request builder. Probably I haven't picked the best name, maybe `BuildRequestError` or `RequestBuilderError` could work better?

Comment (pierotibou) [reply trace-utils/src/send_data.rs:23]: Maybe a comment on the same line will do the work. Don't worry about naming with me, I'm very bad at naming

File: trace-utils/src/send_data.rs
Comment (hoolioh) [reply trace-utils/src/send_data.rs:252]: Nice catch.

File: trace-utils/src/send_data.rs
Comment (pierotibou) [reply trace-utils/src/send_data.rs:106]: I don't know if there's a preferred format. Either what you suggest or something more verbose.
`Status code: <status_code> - Message: <body>`? 
I don't have a real preference so as you wish.

File: trace-utils/src/send_data.rs
Comment (hoolioh) [reply trace-utils/src/send_data.rs]: Since I needed to address some comments I've done this too. Besides it would've taken me more effort to just create the ticket than fixing it.

File: sidecar/src/self_telemetry.rs
Comment (hoolioh) [reply sidecar/src/self_telemetry.rs]: I've commented the code and created a new ticket to track the issue.

File: sidecar/src/self_telemetry.rs
Comment (bantonsson) [inline sidecar/src/self_telemetry.rs:33]: They are both named `trace_chunks_*` in the docs.

File: sidecar/src/self_telemetry.rs
Comment (bantonsson) [reply sidecar/src/self_telemetry.rs:125]: I see that I'm late to the game...

File: sidecar/src/self_telemetry.rs
Comment (bantonsson) [reply sidecar/src/self_telemetry.rs:82]: I'm not too concerned about this part since it's only during reporting.

**Review by pierotibou — APPROVED** [review]
LGTM functionally (so please still wait for another review please). 
Thanks a lot for the implementation and the follow up on all comments.

File: sidecar/src/self_telemetry.rs
Comment (ekump) [inline sidecar/src/self_telemetry.rs:115]: Would it make sense to feature flag this support instead of commenting it out?

File: trace-utils/src/trace_utils.rs
Comment (ekump) [inline trace-utils/src/trace_utils.rs:52]: If we're making things public we should probably include doc comments.

File: sidecar/src/self_telemetry.rs
Comment (ekump) [inline sidecar/src/self_telemetry.rs:73]: I would be remiss if I didn't call out all of these unwraps on `Tag::new()`. If the `Result` is not Ok, should we log an error and not attempt to push the future? I assume we could wrap that logic in a helper function.

Comment (ekump) [reply sidecar/src/self_telemetry.rs:73]: Actually, [this](https://github.com/DataDog/libdatadog/pull/450#issuecomment-2130221322) would solve the problem. So feel free to leave the unwraps for now.

File: trace-utils/src/trace_utils.rs
Comment (hoolioh) [reply trace-utils/src/trace_utils.rs:52]: I got it wrong, it's used only inside the crate so I'm changing to `pub(crate)`.

File: trace-utils/src/send_data.rs
Comment (ekump) [inline trace-utils/src/send_data.rs:10]: probably want to remove this

File: trace-utils/src/send_data.rs
Comment (ekump) [inline trace-utils/src/send_data.rs:386]: DId you mean to remove these comments?

File: trace-utils/src/send_data.rs
Comment (ekump) [inline trace-utils/src/send_data.rs:621]: you could probably just create a `get_tracer_header_tags()` function or implement a default trait and add a `#[cfg(test)]` to it to avoid having to create one in every test.

File: trace-utils/src/send_data.rs
Comment (ekump) [inline trace-utils/src/send_data.rs:49]: These should probably be `///` instead of `//` so they're treated as doc comments?

File: trace-utils/src/send_data.rs
Comment (ekump) [inline trace-utils/src/send_data.rs:115]: This example is a bit misleading since the `.await` is missing from the call to update. You could put the example code within an async function and add the miri ignore so it compiles and doc tests passs

/// #[cfg_attr(miri, ignore)]
/// async fn test_with_poll() {
///    ...
///}

File: trace-utils/src/send_data.rs
Comment (hoolioh) [reply trace-utils/src/send_data.rs:386]: Nope, a casualty from the rebase. Putting it back.

File: trace-utils/src/send_data.rs
Comment (hoolioh) [reply trace-utils/src/send_data.rs:621]: Good catch, I didn't realize I was cluttering the tests when I could use the default impl.

**Review by ekump — APPROVED** [review]

**Pull request merged**
